text/0000-typed-array-conveniences.md

```diff
@@ -0,0 +1,312 @@
+| `JsTypedArray<i32>` | `JsInt32Array`     | [`Int32Array`][Int32Array]         |
+| `JsTypedArray<u64>` | `JsBigUint64Array` | [`BigUint64Array`][BigUint64Array] |
+| `JsTypedArray<i64>` | `JsBigInt64Array`  | [`BigInt64Array`][BigInt64Array]   |
+| `JsTypedArray<f32>` | `JsFloat32Array`   | [`Float32Array`][Float32Array]     |
+| `JsTypedArray<f64>` | `JsFloat64Array`   | [`Float64Array`][Float64Array]     |
+
+These can be convenient to type, and also create a discoverable place for documentation and examples in the API docs.
+
+## Constructing typed arrays
+
+This RFC defines a number of APIs for constructing typed arrays.
+
+### `JsTypedArray`
+
+The `JsTypedArray` type gets three new constructor methods for creating a new typed array.
+
+```rust
+impl<T: Binary> JsTypedArray<T> {
+    // Constructs a new typed array from a length, allocating a new backing buffer.
+    pub fn new<'cx, C>(
+        cx: &mut C,
+        len: usize,
+    ) -> JsResult<'cx, Self>
+    where
+        C: Context<'cx>;
+
+    // Constructs a new typed array from an existing buffer.
+    pub fn from_buffer<'cx, 'b: 'cx, C>(
+        cx: &mut C,
+        buffer: Handle<'b, JsArrayBuffer>,
+    ) -> JsResult<'cx, Self>
+    where
+        C: Context<'cx>;
+
+    // Constructs a new typed array from a region of an existing buffer.
+    pub fn from_region<'c, 'r, C>(
+        cx: &mut C,
+        region: Region<'r, T>,
+    ) -> JsResult<'c, Self>
+    where
+        C: Context<'c>;
+}
+```
+
+### `JsArrayBuffer`
+
+The `JsArrayBuffer` type gets a new method for constructing a `Region` struct, which describes a region of a buffer (see below).
+
+```rust
+impl Handle<'cx, JsArrayBuffer> {
+    pub fn region<T: Binary>(self, byte_offset: usize, len: usize) -> Region<'cx, T>;
+}
+```
+
+### Helper types
+
+#### `neon::types::binary::Binary`
+
+The constructor functions require passing a type tag to Node-API functions, so this RFC defines a sealed `Binary` trait to define on all the primitive element types of typed arrays (`i8`, `u8`, `i16`, `u16`, etc) in order to define the type tag as an associated constant. This is exposed for documentation and to allow userland generic helper functions that work on all typed array types.
+
+```rust
+trait Binary: Sealed + Copy + Debug {
+    const TYPE_TAG: /* hidden */;
+}
+```
+
+#### `neon::types::binary::Region`
+
+The `Region` struct defines a region of an underlying buffer. For performance, it's not actually checked to be a valid region until actually constructing a typed array object. (Otherwise, every method on the type would have to re-validate every time it's called, because of the possibility of the buffer getting detached.)
+
+```rust
+#[derive(Clone, Copy)]
+struct Region<'cx, T: Binary> { /* hidden */ }
+
+impl<'cx, T: Binary> Region<'cx, T> {
+    // Constructs a new typed array for this region.
+    pub fn to_typed_array<'c, C>(self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
```

Comment:
```suggestion
    pub fn to_typed_array<'c, C>(&self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
```

Methods prefix `to_` should take `&self`. Methods that take `self` should be named `into`. Since we're not really consuming anything, I think `&self` and keeping the method makes more sense.

Reply:
Makes sense. It's a bit of a theoretical distinction here since `Region` is `Copy`, but that might still be confusing from a documentation perspective, so I like your suggestion.